### ISSUE TITLE
Video does not seamlessly loop when ControlsApi.getRepeat() is true

### ISSUE THREAD

**Issue by Blir** [caprica/vlcj#892]
I recently upgraded from VLCJ 3.0 t.o 4.0. After initialization, I set:

`mediaPlayerComponent.mediaPlayer().controls().setRepeat(true);`

When the video loops it will briefly show a black screen before it begins the loop again, so the black screen repeatedly appears between each iteration in the loop. This is especially bad for videos that are very short in duration and intended to be viewed as a loop.

In 3.0 I do not believe it behaved this way - although I think I updated VLC at the same time, so maybe this is caused by VLC. I am running VLC 3.0.7.1 and VLCJ 4.2.0.

Is it possible to get this to loop seamlessly (no black screen between loops)?

Thanks for making VLCJ. It is great.

**Comment by caprica** [comment]
This is not behaviour that has changed from vlcj 3.x to vlcj 4.x, at least not that I know of.

It *may* be possible to achieve what you want by using a `MediaListPlayer` and a `MediaList`, but no guarantees.

You also may be able to achieve what you want by setting the VLC option (use vlc -H in a command-window to see if you can find it) that keep the last frame of the video when it ends, but I suspect even this will cause a noticeable pause before the video plays again.

So, short version, I think you will be out of luck, the code related to repeating media has not materially changed since vlcj 3.x and vlcj 4.x as far as I know.

**Comment by caprica** [comment]
If none of that works, I think I remember someone else posted a similar issue and they ended up having two separate media players an alternating between them - each time media stopped the other media player would start and the first media player would pre-prepare the next video and so on.

**Comment by Blir** [comment]
I tried the `MediaListPlayer` and still had the black screen between loops. Currently trying to get the separate media players approach to work, however my app is now crashing and there are no errors in the console. It does seem to be happening when I call `play()` or `prepare()` on one of the two players. I will try to do some more investigation and follow up. I am instantiating two instances of `EmbeddedMediaListPlayerComponent`, is that what you are suggesting or am I misunderstanding?

**Comment by caprica** [comment]
I'm saying two media player components yes, but they don't have to be the list-player variety - that's up to you.

Rather than me rehash everything here, have a look at #760 hopefully it will help.

**Comment by caprica** [comment]
The problem in #760 is not exactly your use-case, but it is quite similar.

**Comment by Blir** [comment]
I took a look at the VLC arguments and the only one I found that is like you described is `--play-and-pause`, but it actually pauses the video so I do not think that will work.

I am trying the CardLayout approach now. I am having trouble getting it to loop at all this way, my code seems to be hanging on execution after invoking `ControlsApi.play()`. I've gotten it to loop once but there was still a black screen, so I'm suspecting the second media player isn't ready in time. Will continue trying to get it to work.

**Comment by caprica** [comment]
What do you mean a "hang"?

Make sure you're not doing anything "clever" with threads.

**Comment by caprica** [comment]
And the thing about the second media player not being ready, you have all the time the first media player is still playing to prepare that second media player. It takes barely any time to invoke mp.media().prepare(mrl);

As soon as that first media player finished, it should be taking the next media from the queue and preparing it immediately - all while the other media player has started playing.

I'm not saying it's *easy* to do this, but I see no reason it shouldn't work.

And just to reiterate, make sure you are doing everything on the proper thread.

**Comment by Blir** [comment]
It wasn't proceeding to the next line of code after the `ControlsApi.play()` invocation. However I found the problem, I forgot to wrap the action in a `MediaPlayer.submit()` invocation, it is looping properly now but I'm still seeing a black screen. Here's the general outline of what I'm currently doing:

`    private final MediaEventListener mediaEventListener = new MediaEventAdapter() {
    	@Override
        public void mediaStateChanged(Media media, State newState) {
    		if (newState == State.ENDED) {
    			player().mediaPlayer().submit(() -> {
    		    	cardLayout.next(cardPanel);
    		    	setNextPlayer();
    		    	player().mediaPlayer().controls().play();
    		    	nextPlayer().mediaPlayer().media().prepare(path);
    			});
    		}
        }
    };`

**Comment by caprica** [comment]
Remember above where I said to make sure you are doing everything on the proper thread? :)

So first...

 - don't go off "state changed", listen for "finished", "error", "playing", "stopped" etc
 - these events are running on a native thread
   - you are correctly using submit() for your media player invocation
   - however you are incorrectly trying to change UI state on a native event thread, you must use SwingUtilities.invokeLater (for Swing) or Platform.runLater(for JavaFX)

And as you found, yep, you also initially hit a classic deadlock by trying to reenter the media player on the same native thread that is processing the event - but you found a solution for that.

**Comment by caprica** [comment]
This is something that changed with vlcj 4.x, there is an explanation on the project README page about it - look on the github project page for "Threading Model".

**Comment by Blir** [comment]
> you are incorrectly trying to change UI state on a native event thread

Are you referring to the `cardLayout.next(cardPanel);` and `player().revalidate();` lines? That makes sense, I do often forget to do things the proper way with regards to UI modifications specifically because usually it never causes a difference.

I did read up on the threading model in the past. It's just easy to forget to do it properly from time to time.

I've made those changes but I'm still seeing the same behavior, a brief black screen between loops. I also removed the validate() call because it didn't seem to be doing anything.

```
    private final MediaPlayerEventListener mpListener = new MediaPlayerEventAdapter() {
        @Override
        public void finished(MediaPlayer mediaPlayer) {
        	player().mediaPlayer().submit(() -> {
        	    SwingUtilities.invokeLater(() -> {
            	        cardLayout.next(cardPanel);
        	    });
        	    setNextPlayer();
        	    player().mediaPlayer().controls().play();
        	    nextPlayer().mediaPlayer().media().prepare(path);
                });
        }
    };

**Comment by caprica** [comment]
Any method invocation that affects UI state must be made on the appropriate toolkit thread. So yep, that means changing the card pane, invalidating layouts and so on.

In your latest example, move the invokeLater to be outside and before the submit call.

You may need to do some tweaking, such as swapping playback on 99% position rather than waiting for a finished event. And also maybe don't switch to the card pane until you get a "ready" event, this guarantees the position is > 0 before you show the video at the expense of maybe missing a frame or two of video.

The only thing that should be in the submit is the stuff that performs media player operations, on the media player that raised the current event.

You should try playing the next player before the submit too (it should be OK on this thread because its a different media player, I don't know for sure tough).

I know this is approach is not ideal and a little tricksy, but you can probably get something workable out of it. It's going to take a bit of empirical tweaking to get it just right.

In short, put the *minimal* amount of stuff in the submit().

**Comment by Blir** [comment]
It does not look like the 99% position approach will work. I am seeing `MediaPlayerEventListener.positionChanged()` only get called one time during the life time of the video playback with a position of about 0.26 (this particular video is less than a second long).

I believe I am close to something that works by
1. beginning playback of the next player as soon as it is ready, but before I switch the card layout
2. using the `--play-and-pause` command line option for VLC

It will loop once with this method without showing a black screen. However after the first loop I am running into an issue where `MediaPlayerEventListener.mediaPlayerReady()` is no longer getting called. I suspect this is because I am calling `MediaApi.prepare()` with the same value as I did for the first invocation, and so technically it is already ready before I call prepare().

So, I think I need to adjust my approach to not depend on the ready event, perhaps by beginning playback of the second player after a timed offset based on the video duration. I'm looking through the VLCJ API to see if video duration is available but I do not see it so far.

**Comment by caprica** [comment]
You will only get a ready event after the media has started playing. It is triggered on receipt of the first native "position changed" event.

If the position has definitely changed and you do not get a media ready event, that's a bug in vlcj. I'll be surprised if this is the case.

So "ready" means "received first position changed event". It isn't really anything to do with the preparedness of the media.

The ready event is exactly what you need from what you describe in your last paragraph.

And as an aside video duration is available via the mediaPlayer.status() API I think, I can't check right now, but it's definitely somewhere.

**Comment by Blir** [comment]
If I am understanding you right, then I think it may be bugged.

Using the below code with the VLC `--play-and-pause` argument set, "mediaPlayerReady" will only be printed once using the below code despite the media being replayed on every pause.

I've switched to listening to the `playing()` event but it's proving unreliable as sometimes it isn't called until the position in the media is already like 0.84.

```
    private final MediaPlayerEventListener bugTest = new MediaPlayerEventAdapter() {
        
        @Override
        public void paused(MediaPlayer mediaPlayer) {
        	System.out.println("paused");
        	mediaPlayer.submit(() -> {
        		mediaPlayer.media().play(path);
        	});
        }
    	
        @Override
        public void mediaPlayerReady(MediaPlayer mediaPlayer) {
        	System.out.println("mediaPlayerReady");
        }
    };

**Comment by Blir** [comment]
I found a simpler approach that is working quite well, using the listener below in addition to the `--play-and-pause`  VLC command line argument. With this method there is no black screen and the stuttering is fairly minimal. So the approach of using two different players was not necessary after all.

Thank you caprica for all your help!

```
	private final MediaPlayerEventListener mpListener = new MediaPlayerEventAdapter() {
        
        @Override
        public void paused(MediaPlayer mediaPlayer) {
        	mediaPlayer.submit(() -> {
        		mediaPlayer.controls().setPosition(0f);
        		mediaPlayer.controls().play();
        	});
        }
    };

**Comment by caprica** [comment]
Just to complete the story, "ready" is unrelated to "pause".

Ready relates to playing the media for the first time, or any time after it was "stopped" or "finished".

When playing media after a pause, you will get a "playing" event (and possibly some buffering events beforehand).

And just to check your solution, are you saying "paused" gets raised if you use the "--play-and-pause" option? If so, that's pretty useful to know and quite a nice trick you've found.

**Comment by Blir** [comment]
> are you saying "paused" gets raised if you use the "--play-and-pause" option?

Yep! That is what happens.

**Comment by P00TAT00** [comment]
Any chance of someone explaining how do I get this to work? I'm kind of a beginner and I've tried to follow what you're doing but I don't really understand the "--play-and-pause" thing nor how listening for pause and playing the video after that will get rid of the black screen.

**Comment by rom4ster** [comment]
Why is this acceptable? If a video is set to repeat. I would expect it to do a TRUE repeat. Not some play pause thing i can implement myself with some basic event listeners. This should be a bug and should be treated as such.

**Comment by caprica** [comment]
This behaviour is because of how VLC itself works, there's no bug in vlcj that can be fixed here.

The other comments in this issue are suggesting workarounds that might be worth trying.

If you feel so strongly that this is unacceptable, then go look in the VLC bug tracker for this issue, provide a fix for it, or fund someone else to do so.

**Comment by caprica** [comment]
For the latest on this see #1227 which has another potential workaround, but is still sub-optimal.

**Comment by caprica** [comment]
Maybe #1243 will provide a viable solution for this.
